src/v/utils/stream_utils.h

```diff
@@ -0,0 +1,251 @@
+    /// \param read_ahead is a number of buffers that can be in-flight at the
+    /// same time
+    ///        (should be relatively low, e.g. 10)
+    /// \param max_size is an optional size limit for in-flight data
+    input_stream_fanout(
+      ss::input_stream<Ch> i,
+      size_t num_clients,
+      size_t read_ahead,
+      std::optional<size_t> max_size)
+      : _num_clients(num_clients)
+      , _bitmask(0)
+      , _max_size(max_size ? *max_size / read_ahead : 0)
+      , _in(std::move(i))
+      , _sem(read_ahead) {
+        vassert(
+          num_clients <= 10 && num_clients >= 2,
+          "input_stream_fanout can have up to 10 clients, {} were given",
+          num_clients);
+        vassert(
+          read_ahead != 0,
+          "input_stream_fanaout can't have zero readahead value");
+    }
+
+    /// Start consuming the input stream
+    void start() {
+        // Run produce in the background until stopped
+        // or produced everything.
+        (void)produce();
+    }
+
+    /// Stop consuming the input stream
+    ss::future<> stop() {
+        if (_as.abort_requested()) {
+            co_return;
+        }
+        _as.request_abort();
+        _pcond.broadcast();
+        co_await _gate.close();
+    }
+
+    /// Detach client with 'index' from the fanout
+    ///
+    /// The remaining clients can proceed as usual.
+    /// If the client is the last one, the stop method will be invoked.
+    ss::future<> detach(unsigned index) {
+        auto m = 1U << index;
+        _bitmask |= m;
+        if (_bitmask == (1U << _num_clients) - 1) {
+            // This is the last client, call stop
+            co_await stop();
+        }
+        // update all in-flight buffers
+        for (auto& it : _buffer) {
+            it.mask |= m;
+        }
+    }
+
+    /// Get next buffer from original imput stream
+    ///
+    /// \param index is an index of the consumer
+    /// \return buffer
+    ss::future<ss::temporary_buffer<Ch>> get(unsigned index) {
+        while (!_as.abort_requested()) {
+            auto gen = _cnt;
+            if (auto ob = co_await maybe_get(index); ob.has_value()) {
+                co_return std::move(ob.value());
+            }
+            // We need to wait for the data in the following scenarios:
+            // - The consumer is outrunning other consummers and the producer.
+            // - There is not data in the buffer because consumers are faster
+            // then the producer.
+            try {
+                co_await _pcond.wait([this, gen] { return _cnt != gen; });
+            } catch (const ss::broken_condition_variable&) {
+            }
+        }
+        co_return ss::temporary_buffer<Ch>();
+    }
+
+private:
+    /// Get next buffer from original imput stream
+    ///
+    /// \return buffer or std::nullopt if the consumer need to repeat
+    ss::future<std::optional<ss::temporary_buffer<Ch>>>
+    maybe_get(unsigned index) {
```

Comment:
This function isn't async, it makes it a little harder to verify that `get` doesn't need a gate.